pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -120,13 +130,36 @@ func (t *Terminator) Evict(pods []*v1.Pod) {
 	// b. non-critical daemonsets
 	// c. critical non-daemonsets
 	// d. critical daemonsets
-	if len(nonCriticalNonDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalNonDaemon...)
-	} else if len(nonCriticalDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalDaemon...)
-	} else if len(criticalNonDaemon) != 0 {
-		t.evictionQueue.Add(criticalNonDaemon...)
-	} else if len(criticalDaemon) != 0 {
-		t.evictionQueue.Add(criticalDaemon...)
+	t.EvictInOrder([][]*v1.Pod{
+		nonCriticalNonDaemon,
+		nonCriticalDaemon,
+		criticalNonDaemon,
+		criticalDaemon,
+	})
+}
+
+func (t *Terminator) EvictInOrder(pods [][]*v1.Pod) {
+	for _, podList := range pods {
+		if len(podList) > 0 {
+			// evict the first list of pods that is not empty, ignore the rest
+			t.evictionQueue.Add(nil, podList...)
+			return
+		}
+	}
+}
+
+// if a pod should be deleted to give it the full terminationGracePeriodSeconds of time before the node will shut down, return the time the pod should be deleted
+func (t *Terminator) shouldDeletePodToEnsureGracePeriod(nodeGracePeriodExpirationTime *time.Time, pod *v1.Pod) *time.Time {
+	// check if pod is already terminating
+	if !pod.DeletionTimestamp.IsZero() {
+		return nil
+	}
+	if nodeGracePeriodExpirationTime == nil || pod.Spec.TerminationGracePeriodSeconds == nil {
```

Comment:
If the terminationGracePeriodSeconds isn't set, is this equivalent to us just deleting the pod at the nodeGracePeriodExpirationTime? I'm wondering if we can just treat this value as 0 to collapse this

Reply:
I was thinking that avoiding explicit pod deletes at the time the node was scheduled to be forcefully terminated was a good idea, since it's one less thing hitting the k8s API. Deleting a pod explicitly at the same time the node is being torn down is unlikely to result in any real improvement for handling the graceful termination of the pod.

I don't think there would be any real problems caused by treating TGPS as 0 if it's undefined beyond some extra pod delete calls on the K8s API though.

Reply:
> Deleting a pod explicitly at the same time the node is being torn down is unlikely to result in any real improvement for handling the graceful termination of the pod

So what happens if we don't do the pod deletes? Does that cause the pods to leak? Or does the graceful node shutdown save us from this happening?

Reply:
Applying the out-of-service NoExecute taint deletes the pods. Karpenter is still blocked on node deletion in the finalizer as long as those pods still exist. Pod leakage should not be possible.